tvm_vm/src/executor/zk.rs

```diff
@@ -0,0 +1,981 @@
+    }
+
+    let g1: G1Affine =
+        G1Projective::new_unchecked((&s[0]).into(), (&s[1]).into(), (&s[2]).into()).into();
+
+    if !g1.is_on_curve() || !g1.is_in_correct_subgroup_assuming_on_curve() {
+        return Err(ZkCryptoError::InvalidInput);
+    }
+
+    Ok(g1)
+}
+
+/// Deserialize a G2 projective point from the BN254 construction serialized as
+/// a vector of three vectors each being a vector of two strings into an affine
+/// G2 point in arkworks format. Return an error if the input is not a vector of
+/// the right format or if any of the strings cannot be parsed as a field
+/// element.
+pub(crate) fn g2_affine_from_str_projective(s: &CircomG2) -> Result<G2Affine, ZkCryptoError> {
+    if s.len() != 3 || s[0].len() != 2 || s[1].len() != 2 || s[2].len() != 2 {
+        return Err(ZkCryptoError::InvalidInput);
+    }
+
+    let g2: G2Affine = G2Projective::new_unchecked(
+        Fq2::new((&s[0][0]).into(), (&s[0][1]).into()),
+        Fq2::new((&s[1][0]).into(), (&s[1][1]).into()),
+        Fq2::new((&s[2][0]).into(), (&s[2][1]).into()),
+    )
+    .into();
+
+    if !g2.is_on_curve() || !g2.is_in_correct_subgroup_assuming_on_curve() {
+        return Err(ZkCryptoError::InvalidInput);
+    }
+
+    Ok(g2)
+}
+
+/// A Groth16 proof in the BN254 construction. Thin wrapper around
+/// `ark_groth16::Proof::<ark_bn254::Bn254>`.
+#[derive(Debug, From)]
+pub struct ProofWrapper(pub(crate) ark_groth16::Proof<ark_bn254::Bn254>);
+
+impl ProofWrapper {
+    /// Deserialize a serialized Groth16 proof using arkworks' canonical serialisation format: https://docs.rs/ark-serialize/latest/ark_serialize/.
+    pub fn deserialize(proof_points_as_bytes: &[u8]) -> ZkCryptoResult<Self> {
+        ark_groth16::Proof::<Bn254>::deserialize_compressed(proof_points_as_bytes)
+            .map_err(|_| ZkCryptoError::InvalidInput)
+            .map(ProofWrapper)
+    }
+}
+
+/// Here there are third party zk login Groth16 verification keys taken  for now
+/// for tests todo: will be replaced by our keys later
+/// todo: move all key data to json config file (?), use hash as id
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////
+// static GLOBAL_VERIFYING_KEY: Lazy<PreparedVerifyingKey<Bn254>> = Lazy::new(global_pvk);
+
+/// Corresponding to proofs generated from prover-dev. Used in devnet/testnet.
+// static INSECURE_VERIFYING_KEY: Lazy<PreparedVerifyingKey<Bn254>> = Lazy::new(insecure_pvk);
+
+// static ZKP_VERIFYING_KEYS: Lazy<HashMap<u32, PreparedVerifyingKey<Bn254>>> = Lazy::new(keys);
+
+// todo: will contain our keys later, key ould be a hash of verification key
+// fn keys() -> HashMap<u32, PreparedVerifyingKey<Bn254>> {
+// let mut h = HashMap::new();
+// h.insert(0, insecure_pvk());
+// h.insert(1, global_pvk());
+// h
+// }
+
+/// Load a fixed verifying key from zkLogin.vkey output. This is based on a
+/// local setup and should not use in production.
```

Comment:
This `should not be used in production`?

Reply:
I removed provocative comments around groth16 key pairs, but I still gonna leave these key pairs, for now. They are all only for tests and will not be used in production. Real production key pairs we will get only after conducting proper Powers of Tau ceremony. It will happen soon, but now tomorrow. And we need some keys just for nearest test purposes.